Async support

dinosaure opened the pull request.

dinosaure committed:
First version of ocaml-tls on async

dinosaure committed:
Implement high level interface

dinosaure committed:
Add tracing, loop on reader and writer and return Ok on the reader

dinosaure committed:
A multiple chain with single certificate is a single and clean code

dinosaure committed:
Remove ppx_sexp_conv constraint

dinosaure committed:
Use the async implementation of nocrypto-entropy

dinosaure committed:
Revert "Remove ppx_sexp_conv constraint"

This reverts commit 0b41d63a43b758841ae2cb1c5ca3d1465da6e665.

seliopou commented on async/lib/x509_async.mli:
```diff
@@ -0,0 +1,30 @@
+type authenticator = X509.Authenticator.a
+
+(** [private_of_pems ~cert ~priv_key] is [priv], after reading the private key
+   and certificate chain fro mthe given PEM-encoded files. *)
```
Typo "fro mthe" -> "from the"

seliopou commented on async/lib/x509_async.mli:
```diff
@@ -0,0 +1,30 @@
+  | `Ca_dir of Fpath.t
+  | `Key_fingerprints of Nocrypto.Hash.hash * (string * Cstruct.t) list
+  | `Hex_key_fingerprints of Nocrypto.Hash.hash * (string * string) list
+  | `No_authentication ] -> (authenticator, [ `Msg of string ]) result Deferred.t
```
Newline might make the return type clearer.

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+
+let load_file path =
+  (* XXX(dinosaure): why [Monitor]? It seems the safer way to load a file
+     without exception leaks. *)
```
Yep. It's OK to leave this comment but it's overkill.

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+     without exception leaks. *)
+  Monitor.try_with ~run:`Now
+    (fun () -> Reader.file_contents (Fpath.to_string path) >>| Cstruct.of_string)
+  >>= function
```
The `>>=` could be replaced with a `>>|` and the `return`s in each branch can be deleted.

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+    (fun () -> Reader.file_contents (Fpath.to_string path) >>| Cstruct.of_string)
+  >>= function
+  | Ok _ as v -> return v
+  | Error exn -> return (Rresult.R.error_msgf "Got an error when we load %a: %s" Fpath.pp path (Exn.to_string exn))
```
A more idiomatic way to do this is via the `Or_error` module. Specifically `Or_error.error` might be what you're looking for:

```ocaml
Or_error.error "failed to load file" exn Exn.sexp_of_t
```

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+
+let private_of_pems ~cert ~priv_key =
+  let open X509.Encoding.Pem in
+  load_file cert >>| Rresult.R.map Certificate.of_pem_cstruct >>= fun certs ->
```
You could just use `Result.map` here, and below, since Core is opened in this module.

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+  | Ok certs, Ok (`RSA pk) -> return (Rresult.R.ok (certs, pk))
+  | (Error _ as err), Ok _ -> return err
+  | Ok _, (Error _ as err) -> return err
+  | Error (`Msg err0), Error (`Msg err1) -> assert false (* end of the world! *)
```
I think here the error messages should be combined and returned. There's probably something in `Or_error` or the `Error` module in base that will help you do this.

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+        Fmt.epr "Silently got an error when we tried to load %a: %s" Fpath.pp path err; [])
+
+let authenticator meth =
+  let now = Ptime_clock.now () in
```
This should use `Synchronous_time_source.now ()` and converting the value appropriately. You can get a `Synchronous_time_source.t` by calling `Synchronous_time_source.wall_clock ()`.

seliopou commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,60 @@
+  | `Key_fingerprints (hash, fps) -> return (Rresult.R.ok (fingerprint hash fps))
+  | `Hex_key_fingerprints (hash, fps) ->
+    let fps = List.map ~f:(fun (n, v) -> (n, dotted_hex_to_cs v)) fps in
+    return (Rresult.R.ok (fingerprint hash fps))
```
I think you can get away with just using `Ok` here.

seliopou commented on async/lib/jbuild:
```diff
@@ -0,0 +1,9 @@
+(library
+ ((name tls_async)
+  (modules (tls_async))
+  (libraries (nocrypto.async core async rresult tls))))
```
`rresult` is unnecessary here.

seliopou commented on async/lib/jbuild:
```diff
@@ -0,0 +1,9 @@
+(library
+ ((name x509_async)
+  (modules (x509_async))
+  (libraries (ptime.clock.os core async fpath rresult fmt nocrypto x509))))
```
It's OK to depend on `ptime` but not an OS-specific clock implementation. For that you should use `Synchronous_time_source` as described below.

avsm commented on async/bin/thales.ml:
```diff
@@ -0,0 +1,373 @@
+  | `AES_256_CBC_SHA256
+  | `AES_256_CCM
+  | `AES_256_GCM_SHA384
+  | `_3DES_EDE_CBC_SHA
```
I noticed this due to the odd type name, and then wondered if we shouldn't be disabling the 3DES ciphers by default due to their insecurity.  See [here](https://security.stackexchange.com/questions/103992/tls-rsa-with-3des-ede-cbc-sha-reported-as-112-bits) for a discussion about this particular cipher combination and its effective entropy.  If most clients in the wild support AES ciphers, then perhaps we can drop the older ones from async-tls.

dinosaure committed:
Merge remote-tracking branch 'upstream/master' into async

dinosaure committed:
ocamlformat pass

dinosaure committed:
First pass after @seliopou's review (most about deletion of Rresult package on libraries)

seliopou commented on async/lib/tls_async.ml:
```diff
@@ -0,0 +1,393 @@
+[@@@warning "-32"]
+
+(* This really belongs just about anywhere else: generic unix name resolution. *)
+let resolve host service =
```
This appears to be unused.

seliopou reviewed (commented):
Ok I'm back on review. Sorry it's taken so long.

As a general comment, there are some functions in the `tls_async.ml` that raise when really they could just return a `Deferred.Or_error.t`. That would make "exception-raising" behavior of the functions more explicit and is a bit more idiomatic. More coming but I thought I'd point this out as I continue on with more specific comments.

seliopou commented on async/lib/tls_async.ml:
```diff
@@ -0,0 +1,393 @@
+  | [] -> return (error_msgf "No address for %s:%s" host service)
+  | ai :: _ -> return (Ok ai.ai_addr)
+
+let gettimeofday = Unix.gettimeofday
```
This is also unused.

seliopou commented on async/lib/tls_async.ml:
```diff
@@ -0,0 +1,393 @@
+    )
+    >>= function Ok res -> return res | Error exn -> return (Error exn)
+
+  let bad_fd () = assert false
```
This should be an informative error message.

seliopou commented:
As a concrete example of a function that should return `Deferred.Or_error.t`, look at `reneg`. I imagine others could benefit from this change as well.

seliopou commented:
Another piece of general feedback, there are a lot of places where `assert false` is being used, I believe, to just raise an error. See the TODO in `reneg` and the `assert false` in `accept`. These should all be replaced with informative error messages. If an `assert false` is used to indicate an impossible case, that's fine, but a comment as to why should also accompany it. I see that this approach was taken in `epoch`, which is great!

seliopou commented:
Another piece of general feedback, there are a lot of places where `assert false` is being used, I believe, to just raise an error. See the TODO in `reneg` and the `assert false` in `accept`. These should all be replaced with informative error messages. If an `assert false` is used to indicate an impossible case, that's fine, but a comment as to why should also accompany it.

seliopou commented on async/lib/tls_async.ml:
```diff
@@ -0,0 +1,393 @@
+        cstruct_to_bigstring buf
+        @@ fun buf ~pos ~len ->
+        Unix.Syscall_result.Int.ok_or_unix_error_exn ~syscall_name:"read"
+        @@ Bigstring.read_assume_fd_is_nonblocking fd buf ~pos ~len
```
Stylistic, but this `@@` style is a kindof unreadable.

seliopou commented on async/lib/tls_async.ml:
```diff
@@ -0,0 +1,393 @@
+  ( recording_errors Async_cstruct.reader_from_socket
+  , recording_errors Async_cstruct.writer_from_socket )
+
+let recv_buf = Cstruct.create 4096
```
This should be allocated on a per-connection basis.

dinosaure committed:
Remove useless function in x509_async.ml

dinosaure committed:
Use Core.Or_error.t in x509_async.ml instead result type

dinosaure committed:
Remove useless function in tls_async.ml

dinosaure committed:
Add Bad_fd exception when syscall raise it

dinosaure committed:
Buffer per state instead global buffer in tls_async.ml

dinosaure committed:
Add exception with documentation instead assert false - from lwt pov, we have the same behavior

dinosaure committed:
Update thales to use Core.Or_error.t instead result type

dinosaure commented:
I think, all serious feedbacks was done with the second pass (with more explicit commits). At the end, stylistic issue can be fixed and `_3DES_EDE_CBC_SHA` can be removed. I removed all `assert false` (noticed with a `TODO`). Most of them are replaced by an exception. On the LWT back-end, it's exactly the same - but, instead to have our own exceptions, LWT back-end raises `Invalid_argument`.

dinosaure commented:
And if all are happy with this impl. it could be a gift for Christmas?

hannesm committed:
Tls_lwt: delay error from writing to peer while reading.

previously, if handle_tls returned a response to the peer, write was called,
which may error out (recording it's error in the state and Lwt.fail). the
received data was never returned to the caller.

now, similar to the semantics in tls_mirage, the error from write is only
recorded in the state, and returned on the next function call (write/..).

hannesm committed:
tls_lwt: record io error only if state used to be active

this ensures proper Lwt_io interoperability, esp. if an Eof has been received,
which in the TLS world is replied to with a close_notify alert -- this write
can certainly fail. earlier, this resulted in an `Error state, now the `Eof
state is retained and Lwt_io can read all application data.

hannesm committed:
upgrade to opam2

hannesm committed:
Merge pull request #389 from hannesm/opam2

upgrade to opam2

hannesm committed:
lib: do not require ciphersuites sent by the client to be a proper set.

this fixes interoperability with (at least) android 4.4 devices

hannesm committed:
Merge pull request #390 from hannesm/no-require-cc-set

lib: do not require the list of ciphersuites sent by the client to be a proper set

hannesm committed:
changes for 0.9.3

hannesm committed:
update test scripts for openssl 1.1.1 (no more 3des and rc4)

dinosaure committed:
Fix typo on Tls_can't_renegotiate

dinosaure committed:
Remove raise of Bad_fd - exception still exists but it returned instead raised

dinosaure committed:
Fix typo on Tls_can't_renegociate

dinosaure committed:
Delete some raise, save exception on the state and do nothing if `Error case appear

dinosaure committed:
Merge remote-tracking branch 'upstream/master' into async

dinosaure committed:
Add a comment to explain diff between async and lwt when we consider mirleft#388

dinosaure committed:
Delete all exceptions in tls_async.ml module

dinosaure committed:
Update thales from the new interface of tls_async.ml

cfcs commented on async/lib/x509_async.ml:
```diff
@@ -0,0 +1,72 @@
+                Fpath.pp path Error.pp err ;
+              [] )
+
+let authenticator meth =
```
- Is there a reason that `now` not just called `time` so it can be applied with `~time` instead of `~time:now` ?
- This reads the current time and does time conversion for all cases, including `` `No_authentication`` where it's not needed.
ping @dinosaure

dinosaure replied:
Hmmhmm, for sure we can update `time` to be `unit -> Ptime.t` and call it then.
